### DISCUSSION TITLE
Redirecting from a button in a table's row

### DISCUSSION THREAD

**Discussion by PoshoDev** [zauberzeug/nicegui#1641] [Q&A] [answered]
### Question

First of all, I'm having a lot of fun with niceGUI so far, it really feels like a gift from heaven, haha. 😁

I'm experimenting a bit with tables, and noticed that adding elements to them is not as straightforward as everything else and will mostly require a bit of HTML scripting. All I'm trying to do at the moment is having a button at the beginning of each row that redirects to a specific link. Looking into Quasar, it seems like normally you'd need to define a redirection function with JavaScript, but I have no clue where to begin with that.

I'm using [this example from the docs](https://nicegui.io/documentation/table#table_with_expandable_rows) as the base for my script.
```python
from nicegui import ui

table = ui.table(columns=columns, rows=rows, row_key='name').classes('w-72')
    table.add_slot('header', r'''
        <q-tr :props="props">
            <q-th auto-width />
            <q-th v-for="col in props.cols" :key="col.name" :props="props">
                {{ col.label }}
            </q-th>
        </q-tr>
    ''')
    table.add_slot('body', r'''
        <q-tr :props="props">
            <q-td auto-width>
                <q-btn size="sm" color="accent" round dense
                    @click="() => $router.push('{{ props.row.link }}')"
                    :icon="'visibility'" />
            </q-td>
            <q-td v-for="col in props.cols" :key="col.name" :props="props">
                {{ col.value }}
            </q-td>
        </q-tr>
    ''')

ui.run()
```

Thank you very much for reading this. Any guidance would be appreciated. :smiling_face_with_tear:

**Comment by falkoschindler** [comment]
Thanks a lot for your warm words, @PoshoDev! 😊

You're approach using template slots is good, but can be simplified. Quasar's QTable allows you to specify [templates for individual cells](https://quasar.dev/vue-components/table#example--body-cell-name-slot), so you don't need to write the whole body template. And maybe you don't need buttons but simple links:

```py
columns = [
    {'name': 'name', 'label': 'Name', 'field': 'name'},
    {'name': 'age', 'label': 'Age', 'field': 'age'},
    {'name': 'link', 'label': 'Link', 'field': 'link'},
]
rows = [
    {'name': 'Alice', 'age': 18, 'link': 'https://example.com'},
    {'name': 'Bob', 'age': 21, 'link': 'https://example.com'},
    {'name': 'Carl', 'age': 42, 'link': 'https://example.com'},
]
table = ui.table(columns=columns, rows=rows, row_key='name')
table.add_slot('body-cell-link', r'<td :props="props"><a :href="props.row.link">{{props.row.link}}</a></td>')
```
<img width="285" alt="Screenshot 2023-09-18 at 07 43 53" src="https://github.com/zauberzeug/nicegui/assets/5767091/7906a46f-b090-4a86-92a4-a81b934765ff">

**Reply by PoshoDev** [reply, marked as answer]
Thank you very much, I actually do need buttons since the table I'm trying to create uses many more columns and having URLs in the table would make it get bloated very quick. However, thanks to your example, I was able to figure out how to do exactly what I needed!

```python
from nicegui import ui

columns = [
        {'name': 'name', 'label': 'Name', 'field': 'name'},
        {'name': 'age', 'label': 'Age', 'field': 'age'},
        {'name': 'link', 'label': 'Link', 'field': 'link'},
    ]
    rows = [
        {'name': 'Alice', 'age': 18, 'link': 'https://example.com'},
        {'name': 'Bob', 'age': 21, 'link': 'https://example.com'},
        {'name': 'Carl', 'age': 42, 'link': 'https://example.com'},
    ]
    table = ui.table(columns=columns, rows=rows, row_key='name')
    table.add_slot('body-cell-link', r'''
                   <td :props="props">
                        <q-btn size="sm" color="accent" round dense 
                        :href="props.row.link" icon="visibility" />
                   </td>
                   ''')

site.run()
```

![Screenshot from 2023-09-18 15-03-33](https://github.com/zauberzeug/nicegui/assets/32964490/86d99c9c-9240-4fbd-97fe-7f57333a30fd)

Thank you so much! Seriously! :smile:

**Reply by retsyo** [reply]
what if every site uses its own icon file? Thanks

**Reply by falkoschindler** [reply]
```py
columns = [
    {'name': 'name', 'label': 'Name', 'field': 'name'},
    {'name': 'age', 'label': 'Age', 'field': 'age'},
    {'name': 'link', 'label': 'Link', 'field': 'link'},
    {'name': 'icon', 'label': 'Icon', 'field': 'icon', 'classes': 'hidden', 'headerClasses': 'hidden'},
]
rows = [
    {'name': 'Alice', 'age': 18, 'link': 'https://example.com', 'icon': 'home'},
    {'name': 'Bob', 'age': 21, 'link': 'https://example.com', 'icon': 'home'},
    {'name': 'Carl', 'age': 42, 'link': 'https://example.com', 'icon': 'home'},
]
table = ui.table(columns=columns, rows=rows, row_key='name')
table.add_slot('body-cell-link', r'''
    <td :props="props">
        <q-btn size="sm" color="accent" round dense :href="props.row.link" :icon="props.row.icon" />
    </td>
''')
```

**Reply by UStar987** [reply]
Hello, how to set up a new page after clicking LINK?

**Reply by rodja** [reply]
@UStar987 how is this related to the current topic? Could you please create a new question? And explain your problem in more detail with a reproducible example. That makes it much easier for the community to help you.
